### DISCUSSION TITLE
Enhancement: Python scripts management and execution in Open WebUI

### DISCUSSION THREAD

**Discussion by powerfooI** [open-webui/open-webui#4165] [General]
# Background

As issue #3583 said, users should be able to upload their python files execute them in Open WebUI.

I've been doing this since several weeks ago and finished python script management features like CRUD, uploading/downloading and code execution. I'll open the pull request if the discussion post goes well. 

# Code Changes

1. Added a table named `script` that stores information of python scripts
2. Added a group of API located at endpoint `/api/v1/scripts` handling CRUD of scripts
3. Added a svelte page containing `Python Playground` in which users could manage and execute their python scripts.
4. Added two properties `readOnly` and `key` for existing component `CodeEditor` for better serving scripts management.

# Description

Due to this is a feature for all user, I put an entry named `Python Playground` in the `UserMenu` no matter which role does current user act as. (figure 1, 2)

There is a sidebar and a editor panel in `Python Playground`. When a user enters python playground, editor panel is empty due to empty script selection. (figure 3)

If the user click button `Create new script`, editor panel appears and is filled with boilerplate python code (a simple quick sort implementation). The user could edit the code, fill the name and description. Click `Create` button and the script is stored by backend. User could read, delete and download its python scripts. (figure 4, 5)

The script list on the left side does not contain content of any script. The content code will be fetched from backend only after a user clicks one item in the list.

Whenever a user is reading or modifying its code, there are two text buttons, `Run` and `Copy`, at the right-top corner of the code editor. Click `Run` and pyodide worker will execute code in current editor panel and show stdout/stderr. (figure 6) Click `Copy` and the code will be copied to user's clipboard. 

Of course a user could upload its python script from local through `Upload new script` button or upload icon button. After file selection and confirming, the uploaded python script would be stored in the backend and showed in the left sidebar. (figure 7, 8)

Python scripts owned by different users are isolated by `user_id`. However, admin could get all scripts. Admin could create its own scripts and its own scripts has a tag noting the belonging. When the admin click a script belonging to others, the title of editor panel will show the owner.

# Figures

### figure 1, 2 - Entry Position

<img width="378" alt="image" src="https://github.com/user-attachments/assets/e2aeabea-5d61-48df-a5f7-ad1fe5fbe831">

<img width="370" alt="image" src="https://github.com/user-attachments/assets/b0e135ac-8bc5-4c10-bd08-c947a676ef19">

### figure 3 - Empty Playground

<img width="1758" alt="image" src="https://github.com/user-attachments/assets/dd004d5f-f2f5-44aa-8e31-5a2ec85119a0">

### figure 4, 5 - Create Script

<img width="1758" alt="image" src="https://github.com/user-attachments/assets/92b4d538-2faa-42f2-a622-c07a9b2c9d04">

<img width="1758" alt="image" src="https://github.com/user-attachments/assets/018dc7ec-994c-450f-9d3d-8711bd52d95f">

### figure 6 - Code Execution

<img width="1758" alt="image" src="https://github.com/user-attachments/assets/d28af155-9e24-4ffe-a823-5259261cf89f">

### figure 7, 8 - Upload Script

<img width="1758" alt="image" src="https://github.com/user-attachments/assets/fc83818d-7dbb-45e7-aabe-3cb97fbfd262">

<img width="1758" alt="image" src="https://github.com/user-attachments/assets/10f0326e-cf7a-412f-bfb4-eff3f4e1d752">

### figure 9 - Admin's View

<img width="1745" alt="image" src="https://github.com/user-attachments/assets/0c115d41-1d8c-4d9f-a48a-da4fec2e2aaf">

**Comment by ricky-davis** [comment]
Can I ask what you see as the *purpose* of this? What would a typical use-case be?

**Reply by powerfooI** [reply]
I think the main purpose of this is to debug `functions` or `tools`. Users are able to import well-debugged code into functions or tools from python playground in the future.

Moreover, users can save code snippets into the playground (as following image shows) and do some modifications and experiments in the future, which is also useful and funny for users.

<img width="1758" alt="image" src="https://github.com/user-attachments/assets/3f01c118-23ac-4047-951a-64063b9d5763">

**Reply by leldr** [reply]
hello! i stumbled here and have a use case in which I think having a "python playground" would be nice.

I have a secured, public facing tiddywiki server that supports API calling. one of the calls returns a JSON of all of my note titles and their respective metadata. I've already written a python script that makes this request using `curl` and pulls my login credentials from environmental variables to successfully make the request.

I'd like to be able to have a conversation with my llama3.2 in open-webui that, when requested, it able to execute this script and include its output into the the query context. I would be surprised if I can't do this with existing Tools & Functions, but I'm currently in the throes of figuring this out. This approach seems straightforward enough for my use case.
